drivers/ieee802154/Kconfig

```diff
@@ -103,6 +103,16 @@ config IEEE802154_SELECTIVE_TXPOWER
 	help
 	  Enable support for selectively setting TX power for every transmission request.
 
+config IEEE802154_MULTIPLE_CCA
+	bool "Support for multiple CCA attempts before transmission"
+	help
+	  This is an optional extension not conforming to IEEE Std. 802.15.4-2015.
```

Comment:
To be honest I see such non-standard extensions that will probably not be implemented by any other driver soon critical in a public API. This is a Nordic-only extension. Such proprietary extensions in the public radio API have been commonplace in the past but this trend [should be stopped](https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1220059230). It should never have come so far.

At the very least a good explanation should been given why such an extension is required and sensible in a cross-vendor API (i.e. useful for the community as a whole) and why it cannot be sensibly done in a "soft MAC" fashion.

Reply:
So maybe an idea would be adding some vendor specific placeholder in `struct net_pkt_cb_ieee802154` ?

Not specifying for what this placeholder would be, this could be decided inside `ieee802154_nrf5.c` for us, but also other vendors could use it in a way they wish inside their drivers.

Reply:
@fgrandel I will open an additional PR moving this feature to ieee802154_nrf5.c, I will provide vendor specific call to set extra_cca_attempts.

Reply:
@ankuns If you like you can also add the preparatory changes as proposed by @ahasztag to this PR, I don't think @rlubos has a problem if we mix up preparatory net/ieee802154 stuff with nRF5 driver stuff in one PR if that makes it easier for you. And maybe it's also a good idea to wait for @rlubos opinion first as he's the net maintainer and I'm just making a proposal. (Just trying to spare you unnecessary effort but of course I'll review it in whatever form you provide it.)

Reply:
@ankuns @rlubos @ahasztag Or even better (just remembering) - we can use the existing `net_pkt->cb` pointer to point to a driver-specific structure that has `struct net_pkt_cb_ieee802154` as its first member and is therefore compatible with it and then just change `net_pkt->cb` to `void *`. This is actually how `net_pkt->cb` was intended in the first place. I just didn't make it `void *` because there was no need. But the need has now come.

Reply:
And to give due credit: https://elixir.bootlin.com/linux/v6.4/source/include/linux/skbuff.h#L745 This pattern was not invented by me. :-)

Reply:
Alternative approach on vendor-level only in this PR https://github.com/zephyrproject-rtos/zephyr/pull/60390

Reply:
@fgrandel as you diagnosed here https://github.com/zephyrproject-rtos/zephyr/pull/60390#issuecomment-1635720164
the alternative https://github.com/zephyrproject-rtos/zephyr/pull/60390 vendor-specific approach is not scalable and also prone to race conditions. 
A field embedded inside `net_pkt` (not directly but in struct net_pkt_cb_ieee802154) allows to use the feature without flaws mentioned https://github.com/zephyrproject-rtos/zephyr/pull/60390#discussion_r1263631588 and https://github.com/zephyrproject-rtos/zephyr/pull/60390#issuecomment-1635701165

I also don't force other vendors to implement it, because the feature must be enabled by `IEEE802154_MULTIPLE_CCA` Kconfig which is disabled by default. Other vendors can only benefit because of commonly defined way to express multiple cca. The feature does not increase memory footprint unless user explicitly enables it. 

When it comes to 
> why it cannot be sensibly done in a "soft MAC" fashion

It possibly could be done on this level however with no predictable time between consecutive CCA attempts.

Reply:
> Other vendors can only benefit because of commonly defined way to express multiple cca.

That remains to be proven - unless another vendor really want's it such a promise is vain IMO.

The configuration part:

Anything in the IEEE802154_ namespace should be standards compliant. That is something I find very hard to give up. If it's a Thread concept then it can be placed in some Kconfig.thread (as an extension of Kconfig.radio or inside the OT module). If it's an nRF5 proprietary extension then it needs to be placed in Kconfig.nrf5. Everything else will clearly break encapsulation IMO.

The data part:

There needs to be a way to include driver and/or OT-specific data inside net_pkt with static allocation, we agree on that. It first has to be decided whether it is OT (which it must be - as they call it!) or nRF5 who owns the new concept.

Assuming it is an OT concept the solution is easy as OT L2 requirement can be naturally exposed in ieee802154_pkt.h (assuming we have a Kconfig constant with appropriate OT-specific prefix). If it is an nRF5 concept, then we need a new approach. As long as no one comes up with a better idea, I'd accept that we introduce driver-specific extensions in `struct net_pkt_cb_ieee802154` in the same way that the decoupling has been done between net_* and ieee802154_*. But then please provide a PR for that and explain why it is nRF and/or OT.

Reply:
> The correct offloading approach in this case would be "scheduled CCA" in the same way we do scheduled TX.

It seems with soft-MAC "scheduled CCA" we can achieve predictable timing, but not attempts as close to each other as possible. With typical asynchronous soft-MAC CCA, we can achieve as close attempts to each other as possible, but with unpredictable timing.

Offloading to the low-level driver allows to achieve both properties simultaneously.

Reply:
Just to add my 2¢, and sorry I skipped all of the additional comments.

@fgrandel Nordic actively uses Zephyr to prototype candidates for inclusion in various standards (e.g. BLE and I would assume 15.4 too). It's pretty incredible actually because it means that Zephyr gets support for the newest BLE features "automagically" (really there is a lot of engineering effort behind it obviously).

Truly, it would be great if we had interoperability with other vendors implementing new features in Zephyr before they became standardized.

In terms of policy, non-standard features shouldn't be enabled by default, and that's in line with Zephyr's policy of default-n selection. I can see non-standard features being they Kconfig-selectable though.

It may make sense, particularly in terms of 15.4, to have a high-level Kconfig for 15.4 to allow experimental / candidate features, because clearly we don't want to be breaking spec without a good reason.

Again, just my 2¢

Reply:
> Truly, it would be great if we had interoperability with other vendors implementing new features in Zephyr before they became standardized.

Yes 100% agreed - that's exactly the reason why I added the word "future" above. It doesn't change any of the remaining argument, though, IMO. Implementing future standard features still requires proper encapsulation at the right level of abstraction which is 100% doable.

Reply:
> It seems with soft-MAC "scheduled CCA" we can achieve predictable timing, but not attempts as close to each other as possible.

@hubertmis CCA duration is well defined so the only challenge is to schedule the next CCA early enough. Soft MAC wouldn't have to wait for the previous CCA to finish before scheduling the next one. That should work with a simple callback/driver event directly after the previous CCA was scheduled while the main CCA call continues to block until the overall process finished. (Remember what we said about callbacks for hardware support w/o breaking encapsulation?)

That said hard-MAC may still be preferable for some vendor - not for timing reasons but for lower power consumption. If avoiding main CPU intervention is honestly the reason why it is done inside the driver (is it really?) then do it, but do it in a way which at least conceptually leaves soft-MAC a well-documented option (w/o implementing it of course).

It's hard to change the way we think, because "it has always been done like that". But I believe there are very good reasons to change our mindset and fight for more vendor-agnostic solutions in the future as far as Zephyr public APIs are concerned. I'm sure it will feel as effortless as the current way of thinking once we got the habit and have the tools in our toolbox.

Therefore I persist: In a vendor-agnostic framework, soft MAC support is mandatory as much as technically possible and hard MAC support optional - not the other way round. Each feature that doesn't have a documented soft-MAC concept is technical debt on the back of the community and contrary to the spirit of a vendor-agnostic framework like Zephyr.

I believe it is now time to discuss this in the technical working group. We have shown that there is practical relevance in our discussion as well as realistic architectural alternatives and I think it should simply be officially decided what the "Zephyr way" is going to be in such a case in the future.

Reply:
> That said hard-MAC may still be preferable for some vendor - not for timing reasons but for lower power consumption.

Timing reasons are still relevant, but I guess we don't need to go now into a discussion about soft- and hard-MAC scheduling timing and how they can be implemented on different platforms.

> Therefore I persist: In a vendor-agnostic framework, soft MAC support is mandatory as much as technically possible and hard MAC support optional - not the other way round.

Do we all understand soft-MAC as Zephyr's 802.15.4 L2? If so I would say that for optional 802.15.4 features, especially for vendor-specific features, hard-MAC support is optional and soft-MAC support is optional. If one optional feature is missing in the soft-MAC and in a platform's hard-MAC, then this feature cannot be used on this platform.

The discussed feature is a very good example because it brings value primarily to OpenThread L2. If the feature was implemented in soft-MAC, OpenThread L2 could not benefit from this. So to meet the requirements, the feature could have been implemented in OpenThread L2 or in hard-MAC. There is little value in implementing and maintaining it in soft-MAC, so I'm reluctant to agree it's mandatory.

Reply:
> Do we all understand soft-MAC as Zephyr's 802.15.4 L2?

@hubertmis  At least I personally don't atm. Things to be shared between OT and native IEEE 802.15.4 could be placed in a shared L2-level module iif we agree that it makes sense. But note: This is not necessary for the feature we're discussing here as it is only required by OT. A soft-MAC solution for multi-CCA would have to be placed into Zephyr's OT platform adaptation layer - the same that hosts radio.c. Of course not necessarily in the same file but the same architectural module.

> hard-MAC support is optional and soft-MAC support is optional

My current understanding is that we have to distinguish between architecture and implementation:
* regarding implementation I agree with you
* regarding architecture I don't: soft MAC must be architecturally possible for all L2-features required by some L2 unless we can ensure that all vendors' chipsets are guaranteed to provide offloading support for the given L2 feature. Hard MAC otoh only becomes mandatory as soon as the first vendor wants it. This is rather obvious as almost all L2 features currently only have a pure "soft MAC" implementation.

We actually already discussed this in a F2F recently and have defined some practical rules [which I've tried to document in our RFC-like discussion](https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1267034840) for reference.